### PR TITLE
[Core] add and implement `VLLM_LOGITS_PROCESSOR_THREADS`

### PR DESCRIPTION
**Pull request by akeshet**
This PR adds an environment variable `VLLM_LOGITS_PROCESSOR_THREADS`.

If set, this will cause VLLM to use a threadpool of the given size to multithread (across sequences in a batch) its calls to the provided logits processors.

This can increase GPU utilization and decrease ITL in cases where batches are large and where logits processors either (a) launch additional CUDA kernels or (b) do significant CPU-bound work while not holding the python GIL, or both.

### CODE REVIEW

**Comment by akeshet** [comment]
Note: this PR is slightly different from the one I have already tested internally. I'd like to test it a bit before its truly ready for review, but I didn't see an option to create this PR as a draft.

**Comment by akeshet** [comment]
> Note: this PR is slightly different from the one I have already tested internally. I'd like to test it a bit before its truly ready for review, but I didn't see an option to create this PR as a draft.

My local tests look good, ready for review!

**Comment by russellb** [comment]
This makes sense to me. Do you happen to have any performance data you can share? Perhaps benchmarking with structured output and xgrammar? It would be nice to show the benefit concretely. There are some scripts that would help automate this -- benchmarks/benchmark_guided.py and benchmarks/benchmark_serving_guided.py.

Also, I wonder about just making this the default behavior instead of requiring it to be enabled via a tunable. Allowing the threadpool to be adjusted seems fine, but if the benefit is noticeable with one of our in-tree logits processors (xgrammar, in particular), then putting it on by default may make sense.

**Comment by akeshet** [comment]
> This makes sense to me. Do you happen to have any performance data you can share?

I can't share my raw numbers or traces, but I can say that we see a roughly 10% improvement in tok/s and req/s when using this feature in combination with our internal logits processor (and with request concurrency of ~50, and threadpool size 50).

> Also, I wonder about just making this the default behavior instead of requiring it to be enabled via a tunable. Allowing the threadpool to be adjusted seems fine, but if the benefit is noticeable with one of our in-tree logits processors (xgrammar, in particular), then putting it on by default may make sense.

I wanted to be conservative and not alter existing behavior. As is often the case with python multithreading, there are cases where I suspect this would hurt rather than help performance. For instance, to get the benefit from this PR with our internal logits processor logic I had to make some tweaks (a well placed cuda sync) to ensure that the logits processors in separate threads don't block eachother via cuda's memcpy lock.

**Comment by russellb** [comment]
> > This makes sense to me. Do you happen to have any performance data you can share?
> 
> I can't share my raw numbers or traces, but I can say that we see a roughly 10% improvement in tok/s and req/s when using this feature in combination with our internal logits processor (and with request concurrency of ~50, and threadpool size 50).
> 
> > Also, I wonder about just making this the default behavior instead of requiring it to be enabled via a tunable. Allowing the threadpool to be adjusted seems fine, but if the benefit is noticeable with one of our in-tree logits processors (xgrammar, in particular), then putting it on by default may make sense.
> 
> I wanted to be conservative and not alter existing behavior. As is often the case with python multithreading, there are cases where I suspect this would hurt rather than help performance. For instance, to get the benefit from this PR with our internal logits processor logic I had to make some tweaks (a well placed cuda sync) to ensure that the logits processors in separate threads don't block eachother via cuda's memcpy lock.

OK, thanks. If we're not able to show how it helps something in-tree, then I agree with having it off by default.

Can you expand on the doc text to add some guidance on when it would be useful?  Something like "useful when using custom logits processors that either (a) launch additional CUDA kernels or (b) do significant CPU-bound work while not holding the python GIL, or both."

**Comment by akeshet** [comment]
> Can you expand on the doc text to add some guidance on when it would be useful? Something like "useful when using custom logits processors that either (a) launch additional CUDA kernels or (b) do significant CPU-bound work while not holding the python GIL, or both."

Done.

**Review by russellb — APPROVED** [review]
lgtm, thanks!

**Review by mgoin — APPROVED** [review]
LGTM given this is limited to within `logits_processor.py` and is disabled by default, thanks

**Comment by alejopaullier96** [comment]
@akeshet is there a minimal example on how to use logits processor with this environment variable?

**Comment by akeshet** [comment]
> @akeshet is there a minimal example on how to use logits processor with this environment variable?

@alejopaullier96 The minimum example would be simply to `export VLLM_LOGS_PROCESSOR_THREADS=$VALUE` prior to launching vllm. But it will really depend on the details of yours logits processor you are using, as to whether this gives any performance benefit. I came to this optimization by examining `nsys` traces of our workload with a custom logits processor.
